src/react/components/modals/Form.tsx

```diff
@@ -0,0 +1,131 @@
+   This file is part of Astarte.
+
+   Copyright 2020 Ispirata Srl
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import React, { useCallback, useRef, useState } from 'react';
+import { Button, Modal, Spinner } from 'react-bootstrap';
+import type { ModalProps } from 'react-bootstrap';
+import metaSchemaDraft04 from 'ajv/lib/refs/json-schema-draft-04.json';
+import JsonSchemaForm from '@rjsf/bootstrap-4';
+import type { ComponentProps } from 'react';
+
+const additionalMetaSchemas = [metaSchemaDraft04];
+
+type JsonSchemaFormProps = ComponentProps<typeof JsonSchemaForm>;
+
+type BoostrapVariant =
+  | 'primary'
+  | 'secondary'
+  | 'success'
+  | 'warning'
+  | 'danger'
+  | 'info'
+  | 'light'
+  | 'dark'
+  | 'link';
+
+const checkContainedValidForm = (element: HTMLElement) => {
+  return !!element.querySelector('form:valid');
+};
+
+interface Props {
+  cancelLabel?: string;
+  confirmLabel?: string;
+  confirmVariant?: BoostrapVariant;
+  initialData?: JsonSchemaFormProps['formData'];
+  isConfirming?: boolean;
+  onCancel: () => void;
+  onConfirm: (formData: any) => void;
+  schema: JsonSchemaFormProps['schema'];
+  size?: ModalProps['size'];
+  title: React.ReactNode;
+  uiSchema?: JsonSchemaFormProps['uiSchema'];
+}
+
+const FormModal = ({
+  cancelLabel = 'Cancel',
+  confirmLabel = 'Confirm',
+  confirmVariant = 'primary',
+  initialData,
+  isConfirming = false,
+  onCancel,
+  onConfirm,
+  schema,
+  size = 'lg',
+  title,
+  uiSchema,
+}: Props): React.ReactElement => {
+  const modalBodyRef = useRef<HTMLDivElement>(null);
+  const [formData, setFormData] = React.useState(initialData || null);
+  const [isValidForm, setIsValidForm] = useState(false);
+
+  const handleJsonSchemaFormSubmit = useCallback(
+    (event: any) => {
+      onConfirm(event.formData);
+    },
+    [onConfirm],
+  );
+
+  const handleJsonSchemaFormChange = useCallback(
+    (event: any) => {
+      const isValid = modalBodyRef.current != null && checkContainedValidForm(modalBodyRef.current);
+      setFormData(event.formData);
+      setIsValidForm(isValid);
+    },
+    [modalBodyRef, setIsValidForm],
+  );
+
+  return (
+    <Modal show centered size={size} onHide={onCancel}>
+      <Modal.Header closeButton>
+        <Modal.Title>{title}</Modal.Title>
+      </Modal.Header>
+      <Modal.Body>
+        <div ref={modalBodyRef}>
+          <JsonSchemaForm
+            schema={schema}
+            additionalMetaSchemas={additionalMetaSchemas}
```

Comment:
this is referenced only once, you can use it here directly without the need of a global const

Reply:
actually, re require jsonschema 4 only because it's used in astarte block definitions. Here we can use draft 6 schemas and drop this additional schema

Reply:
I added jsonschema 4 definitions so that this component can be reused for astarte blocks in a later PR. We might want to do that to extend modals from a few shared components, thus making it simpler to mantain the same visual style throughout the app.
It is defined as a global variable so that `<JsonSchemaForm />` doesn't get re-rendered because of the literal array expression `[metaSchemaDraft04]`